// @slidev/cli/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});










var _chunkC4IMTHZNjs = require('./chunk-C4IMTHZN.js');



var _chunk5LIGFQWWjs = require('./chunk-5LIGFQWW.js');


var _chunkN6L4BUFDjs = require('./chunk-N6L4BUFD.js');

// node/index.ts
_chunkN6L4BUFDjs.init_cjs_shims.call(void 0, );

// node/declare.ts
_chunkN6L4BUFDjs.init_cjs_shims.call(void 0, );













exports.ViteSlidevPlugin = _chunk5LIGFQWWjs.ViteSlidevPlugin; exports.createServer = _chunkC4IMTHZNjs.createServer; exports.createWindiCSSPlugin = _chunk5LIGFQWWjs.createWindiCSSPlugin; exports.getAddonRoots = _chunkC4IMTHZNjs.getAddonRoots; exports.getCLIRoot = _chunkC4IMTHZNjs.getCLIRoot; exports.getClientRoot = _chunkC4IMTHZNjs.getClientRoot; exports.getRoot = _chunkC4IMTHZNjs.getRoot; exports.getThemeRoots = _chunkC4IMTHZNjs.getThemeRoots; exports.getUserRoot = _chunkC4IMTHZNjs.getUserRoot; exports.isPath = _chunkC4IMTHZNjs.isPath; exports.parser = _chunkC4IMTHZNjs.parser; exports.resolveOptions = _chunkC4IMTHZNjs.resolveOptions;
