// @slidev/parser/dist/fs.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});















var _chunkXTGVXNGIjs = require('./chunk-XTGVXNGI.js');

// src/fs.ts
var _fs = require('fs');
var _path = require('path');
async function load(filepath, themeMeta, content) {
  const dir = _path.dirname.call(void 0, filepath);
  const markdown = content != null ? content : await _fs.promises.readFile(filepath, "utf-8");
  const data = _chunkXTGVXNGIjs.parse.call(void 0, markdown, filepath, themeMeta);
  const entries = /* @__PURE__ */ new Set([
    filepath
  ]);
  for (let iSlide = 0; iSlide < data.slides.length; ) {
    const baseSlide = data.slides[iSlide];
    if (!baseSlide.frontmatter.src) {
      iSlide++;
      continue;
    }
    data.slides.splice(iSlide, 1);
    if (baseSlide.frontmatter.hide)
      continue;
    const srcExpression = baseSlide.frontmatter.src;
    const path = _path.resolve.call(void 0, dir, srcExpression);
    const raw = await _fs.promises.readFile(path, "utf-8");
    const subSlides = _chunkXTGVXNGIjs.parse.call(void 0, raw, path, themeMeta);
    for (const [offset, subSlide] of subSlides.slides.entries()) {
      const slide = { ...baseSlide };
      slide.source = {
        filepath: path,
        ...subSlide
      };
      if (offset === 0 && !baseSlide.frontmatter.srcSequence) {
        slide.inline = { ...baseSlide };
        delete slide.inline.frontmatter.src;
        Object.assign(slide, slide.source, { raw: null });
      } else {
        Object.assign(slide, slide.source);
      }
      const baseSlideFrontMatterWithoutSrc = { ...baseSlide.frontmatter };
      delete baseSlideFrontMatterWithoutSrc.src;
      slide.frontmatter = {
        ...subSlide.frontmatter,
        ...baseSlideFrontMatterWithoutSrc,
        srcSequence: `${baseSlide.frontmatter.srcSequence ? `${baseSlide.frontmatter.srcSequence},` : ""}${srcExpression}`
      };
      data.features = _chunkXTGVXNGIjs.mergeFeatureFlags.call(void 0, data.features, _chunkXTGVXNGIjs.detectFeatures.call(void 0, raw));
      entries.add(path);
      data.slides.splice(iSlide + offset, 0, slide);
    }
  }
  for (let iSlide = 0; iSlide < data.slides.length; iSlide++)
    data.slides[iSlide].index = iSlide === 0 ? 0 : 1 + data.slides[iSlide - 1].index;
  data.entries = Array.from(entries);
  return data;
}
async function save(data, filepath) {
  filepath = filepath || data.filepath;
  await _fs.promises.writeFile(filepath, _chunkXTGVXNGIjs.stringify.call(void 0, data), "utf-8");
}
async function saveExternalSlide(slide) {
  await _fs.promises.writeFile(slide.filepath, _chunkXTGVXNGIjs.stringifySlide.call(void 0, slide), "utf-8");
}



















exports.detectFeatures = _chunkXTGVXNGIjs.detectFeatures; exports.filterDisabled = _chunkXTGVXNGIjs.filterDisabled; exports.load = load; exports.mergeFeatureFlags = _chunkXTGVXNGIjs.mergeFeatureFlags; exports.parse = _chunkXTGVXNGIjs.parse; exports.parseAspectRatio = _chunkXTGVXNGIjs.parseAspectRatio; exports.parseRangeString = _chunkXTGVXNGIjs.parseRangeString; exports.parseSlide = _chunkXTGVXNGIjs.parseSlide; exports.prettify = _chunkXTGVXNGIjs.prettify; exports.prettifySlide = _chunkXTGVXNGIjs.prettifySlide; exports.resolveConfig = _chunkXTGVXNGIjs.resolveConfig; exports.resolveFonts = _chunkXTGVXNGIjs.resolveFonts; exports.save = save; exports.saveExternalSlide = saveExternalSlide; exports.scanMonacoModules = _chunkXTGVXNGIjs.scanMonacoModules; exports.stringify = _chunkXTGVXNGIjs.stringify; exports.stringifySlide = _chunkXTGVXNGIjs.stringifySlide; exports.verifyConfig = _chunkXTGVXNGIjs.verifyConfig;
